File: 322. Coin Change/solution.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        # edge cases
        if not amount:
            return 0
        
        if not coins or amount < 0 or amount < min(coins):
            return -1
        
        # dp
        change = [0]*(amount+1)
        coins.sort()
        
        for i in range(1, amount+1):
            minChange = sys.maxsize
            for coin in coins:
                if i - coin < 0:
                    continue
                elif i - coin == 0:
                    minChange = 1
                else:
                    if change[i-coin] != -1:
                        minChange = min(1 + change[i-coin], minChange)
                        
            if minChange == sys.maxsize:
                minChange = -1
            change[i] = minChange
            
        return change[-1]

File: 322. Coin Change/solution.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        Set<Integer> coinSet = new HashSet<Integer>();
        Arrays.sort(coins);
        int[] dp = new int[amount+1];
        
        for (int i = 1; i < amount+1; i++) {
            int maxCoins = Integer.MAX_VALUE;
            for (int coin: coins) {
                if (i - coin >= 0 && dp[i-coin] >= 0) {
                    maxCoins = Math.min(maxCoins, dp[i-coin] + 1);
                } else if (i - coin < 0) {
                    break;
                }
            }
            if (maxCoins == Integer.MAX_VALUE) {
                dp[i] = -1;
            } else {
                dp[i] = maxCoins;
            }
        }
        
        return dp[amount];
    }
}
